File: LeetCode/2492-minimum-score-of-a-path-between-two-cities/2492-minimum-score-of-a-path-between-two-cities.java
import java.util.*;

class Solution {
    int[] parent;
    public int minScore(int n, int[][] roads) {
        parent = new int[n+1];
        Arrays.fill(parent, -1);

        for (int[] r : roads) {
            merge(r[0], r[1]);
        }

        int root = find(1);
        int ans = Integer.MAX_VALUE;

        for (int[] r : roads) {
            if (find(r[0])==root || find(r[1])==root) {
                ans = Math.min(ans, r[2]);
            }
        }

        return ans;
    }

    private int find(int x) {
        if (parent[x] == -1) {
            return x;
        }
        return parent[x] = find(parent[x]);
    }

    private void merge(int a, int b) {
        a = find(a);
        b = find(b);

        if (a==b) {
            return;
        }

        parent[b] = a;
    }
}

File: LeetCode/2492-minimum-score-of-a-path-between-two-cities/2492-minimum-score-of-a-path-between-two-cities.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    struct Node {
        int v, dist;
    };

    int minScore(int n, vector<vector<int>>& roads) {
        vector<vector<Node>> map(n+1);

        for (const vector<int>& r : roads) {
            int u = r[0];
            int v = r[1];
            int d = r[2];

            map[u].push_back({v, d});
            map[v].push_back({u, d});
        }

        int ans = INT_MAX;
        vector<bool> isVisited(n+1, false);
        isVisited[1] = true;

        queue<int> q;
        q.push(1);

        // 갈 수 있는 모든 경로를 탐색하면서 최솟값 갱신
        while (!q.empty()) {
            int cur = q.front(); q.pop();

            for (const Node& nextNode : map[cur]) {
                ans = min(ans, nextNode.dist);

                if (isVisited[nextNode.v]) {
                    continue;
                }

                isVisited[nextNode.v] = true;
                q.push(nextNode.v);
            }
        }

        return ans;
    }
};